lib/backend/model/rule.go

```diff
@@ -37,26 +39,62 @@ type Rule struct {
 	NotICMPCode *int `json:"!icmp_code,omitempty" validate:"omitempty,gte=1,lte=255"`
 
 	SrcTag      string             `json:"src_tag,omitempty" validate:"omitempty,tag"`
-	SrcNet      *net.IPNet         `json:"src_net,omitempty" validate:"omitempty"`
+	SrcNet      IPNets             `json:"src_net,omitempty" validate:"omitempty"`
 	SrcSelector string             `json:"src_selector,omitempty" validate:"omitempty,selector"`
 	SrcPorts    []numorstring.Port `json:"src_ports,omitempty" validate:"omitempty"`
 	DstTag      string             `json:"dst_tag,omitempty" validate:"omitempty,tag"`
 	DstSelector string             `json:"dst_selector,omitempty" validate:"omitempty,selector"`
-	DstNet      *net.IPNet         `json:"dst_net,omitempty" validate:"omitempty"`
+	DstNet      IPNets             `json:"dst_net,omitempty" validate:"omitempty"`
 	DstPorts    []numorstring.Port `json:"dst_ports,omitempty" validate:"omitempty"`
 
 	NotSrcTag      string             `json:"!src_tag,omitempty" validate:"omitempty,tag"`
-	NotSrcNet      *net.IPNet         `json:"!src_net,omitempty" validate:"omitempty"`
+	NotSrcNet      IPNets             `json:"!src_net,omitempty" validate:"omitempty"`
 	NotSrcSelector string             `json:"!src_selector,omitempty" validate:"omitempty,selector"`
 	NotSrcPorts    []numorstring.Port `json:"!src_ports,omitempty" validate:"omitempty"`
 	NotDstTag      string             `json:"!dst_tag,omitempty" validate:"omitempty"`
 	NotDstSelector string             `json:"!dst_selector,omitempty" validate:"omitempty,selector"`
-	NotDstNet      *net.IPNet         `json:"!dst_net,omitempty" validate:"omitempty"`
+	NotDstNet      IPNets             `json:"!dst_net,omitempty" validate:"omitempty"`
 	NotDstPorts    []numorstring.Port `json:"!dst_ports,omitempty" validate:"omitempty"`
 
 	LogPrefix string `json:"log_prefix,omitempty" validate:"omitempty"`
 }
 
+type IPNets []*net.IPNet
+
+var ErrIPNetsTooShort = errors.New("JSON input too short to be a list of IPs")
+
+func (ns *IPNets) UnmarshalJSON(b []byte) error {
+	if len(b) == 0 {
+		return ErrIPNetsTooShort
+	}
+	if b[0] != '[' {
+		var singleIP net.IPNet
+		err := json.Unmarshal(b, &singleIP)
+		if err != nil {
+			return err
+		}
+		*ns = append(*ns, &singleIP)
+		return nil
+	}
+	nsAsSlice := (*[]*net.IPNet)(ns)
+	return json.Unmarshal(b, &nsAsSlice)
+}
+
+func (ns IPNets) MarshalJSON() ([]byte, error) {
```

Comment:
In the YAML you'll get a single entry returned if configure a list with a single value in it.

I think we should be consistent about returning the same type that you configure (single value vs slice), so I think you'd need to store an explicit indication of whether it was a list or single entry rather that just using the list length.